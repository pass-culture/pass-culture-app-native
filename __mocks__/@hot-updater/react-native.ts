import { ReactNode } from 'react'

const getUpdateSource = jest.fn(() => Promise.resolve({ source: 'source' }))

const HotUpdater = jest.fn(() => (app: ReactNode) => app)

export default Object.assign(HotUpdater, {
  getUpdateSource,
})
