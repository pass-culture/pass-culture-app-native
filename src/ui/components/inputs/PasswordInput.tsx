import { t } from '@lingui/macro'
import React, { forwardRef, useState } from 'react'
import {
  NativeSyntheticEvent,
  TextInput as RNTextInput,
  TextInputFocusEventData,
} from 'react-native'
import styled from 'styled-components/native'
import { v4 as uuidv4 } from 'uuid'

import { accessibilityAndTestId } from 'libs/accessibilityAndTestId'
import { styledButton } from 'ui/components/buttons/styledButton'
import { InputLabel } from 'ui/components/InputLabel'
import { InputContainer } from 'ui/components/inputs/InputContainer'
import { LabelContainer } from 'ui/components/inputs/LabelContainer'
import { RequiredLabel } from 'ui/components/inputs/RequiredLabel'
import { Touchable } from 'ui/components/touchable/Touchable'
import { Eye as DefaultEye } from 'ui/svg/icons/Eye'
import { EyeSlash as DefaultEyeSlash } from 'ui/svg/icons/EyeSlash'
import { Spacer } from 'ui/theme'

import { BaseTextInput } from './BaseTextInput'
import { StyledInputContainer } from './StyledInputContainer'
import { getCustomTextInputProps, getRNTextInputProps, TextInputProps } from './types'

const WithRefPasswordInput: React.ForwardRefRenderFunction<RNTextInput, TextInputProps> = (
  props,
  forwardedRef
) => {
  const nativeProps = getRNTextInputProps(props)
  const customProps = getCustomTextInputProps(props)

  const [shouldHidePassword, setShouldHidePassword] = useState(true)
  const [isFocus, setIsFocus] = useState(false)
  const passwordInputID = uuidv4()

  function togglePasswordDisplay() {
    setShouldHidePassword(!shouldHidePassword)
  }

  function onFocus(e: NativeSyntheticEvent<TextInputFocusEventData>) {
    setIsFocus(true)
    props?.onFocus?.(e)
  }

  function onBlur() {
    setIsFocus(false)
  }

  return (
    <InputContainer>
      {!!customProps.label && (
        <React.Fragment>
          <LabelContainer>
            <InputLabel htmlFor={passwordInputID}>{customProps.label}</InputLabel>
            {!!customProps.isRequiredField && <RequiredLabel />}
          </LabelContainer>
          <Spacer.Column numberOfSpaces={2} />
        </React.Fragment>
      )}
      <StyledInputContainer
        isFocus={isFocus}
        isError={customProps.isError}
        isInputDisabled={customProps.disabled}>
        <StyledBaseTextInput
          {...nativeProps}
          nativeID={passwordInputID}
          onFocus={onFocus}
          onBlur={onBlur}
          secureTextEntry={shouldHidePassword}
          ref={forwardedRef}
          aria-describedby={customProps.accessibilityDescribedBy}
          {...accessibilityAndTestId(t`Mot de passe`)}
        />
        <IconTouchableOpacity
          testID="toggle-password-visibility"
          accessibilityLabel={
            shouldHidePassword ? t`Afficher le mot de passe` : t`Cacher le mot de passe`
          }
          onPress={togglePasswordDisplay}>
          {shouldHidePassword ? <EyeSlash /> : <Eye />}
        </IconTouchableOpacity>
      </StyledInputContainer>
    </InputContainer>
  )
}

export const PasswordInput = forwardRef<RNTextInput, TextInputProps>(WithRefPasswordInput)

const StyledBaseTextInput = styled(BaseTextInput)({
  flex: 0.9,
})

const IconTouchableOpacity = styledButton(Touchable)({
  flex: 0.1,
  maxWidth: 60,
  alignItems: 'flex-end',
})

const EyeSlash = styled(DefaultEyeSlash).attrs(({ theme }) => ({
  size: theme.icons.sizes.small,
}))``

const Eye = styled(DefaultEye).attrs(({ theme }) => ({
  size: theme.icons.sizes.small,
}))``
