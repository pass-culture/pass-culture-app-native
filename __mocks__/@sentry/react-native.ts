import * as ActualSentryModule from '@sentry/react-native'

const MockedSentryModule = jest.genMockFromModule<typeof ActualSentryModule>('@sentry/react-native')

module.exports = {
  ...MockedSentryModule,
}
