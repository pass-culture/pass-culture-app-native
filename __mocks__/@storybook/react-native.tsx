import React from 'react'; // @storybook
import { View } from 'react-native';

const configure = jest.fn();
const getStorybookUI = jest.fn().mockImplementation(() => () => <View />);

export { configure, getStorybookUI };
