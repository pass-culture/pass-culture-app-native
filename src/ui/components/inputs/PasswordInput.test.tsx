import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'
import { TextInput as RNTextInput } from 'react-native'

import { PasswordInput } from './PasswordInput'

describe('<PasswordInput />', () => {
  it('should render correctly', () => {
    const instance = render(<PasswordInput />)

    expect(instance).toMatchSnapshot()
  })

  it('should render correctly when password is displayed', () => {
    const instance = render(<PasswordInput />)
    const hiddenSnapshot = instance.toJSON()

    const eyeSlashIcon = instance.getByTestId('eye-slash')
    fireEvent.press(eyeSlashIcon)

    const displayPasswordSnapshot = instance.toJSON()

    expect(displayPasswordSnapshot).toMatchDiffSnapshot(hiddenSnapshot)
  })

  it('should render correctly when password is hidden after it was displayed', () => {
    const instance = render(<PasswordInput />)

    const eyeSlashIcon = instance.getByTestId('eye-slash')
    fireEvent.press(eyeSlashIcon)

    const displayPasswordSnapshot = instance.toJSON()

    const eyeIcon = instance.getByTestId('eye')
    fireEvent.press(eyeIcon)

    const hiddenPasswordSnapshot = instance.toJSON()

    expect(hiddenPasswordSnapshot).toMatchDiffSnapshot(displayPasswordSnapshot)
  })

  it('should render ref correctly', () => {
    const myRef = React.createRef<RNTextInput>()
    render(<PasswordInput ref={myRef} />)

    expect(myRef.current).toBeTruthy()
  })
})
