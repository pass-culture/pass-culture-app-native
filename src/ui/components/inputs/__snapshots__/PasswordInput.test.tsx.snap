// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PasswordInput /> should render correctly 1`] = `
<View
  isFocus={false}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "borderColor": "#c7c7cc",
        "borderRadius": 22,
        "borderStyle": "solid",
        "borderWidth": 1,
        "flexDirection": "row",
        "height": 40,
        "maxWidth": 675,
        "paddingBottom": 4,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 4,
        "shadowColor": "#151515",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.15,
        "shadowRadius": 6,
        "width": "100%",
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder=""
    placeholderTextColor="#626262"
    rejectResponderTermination={true}
    secureTextEntry={true}
    style={
      Array [
        Object {
          "color": "#151515",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "fontFamily": "Montserrat-Regular",
          "fontSize": 15,
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        },
        Object {
          "flexBasis": 0,
          "flexGrow": 0.9,
          "flexShrink": 1,
        },
      ]
    }
    underlineColorAndroid="transparent"
  />
  <View
    accessibilityLabel="Basculer l'affichage du mot de passe"
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flexBasis": 0,
        "flexGrow": 0.1,
        "flexShrink": 1,
        "maxWidth": 60,
        "opacity": 1,
      }
    }
    testID="Basculer l'affichage du mot de passe"
  >
    <View
      height="100%"
      testID="eye-slash"
      width="100%"
    >
      <Text>
        eye-slash-SVG-Mock
      </Text>
    </View>
  </View>
</View>
`;

exports[`<PasswordInput /> should render correctly when password is displayed 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -33,11 +33,11 @@
      onBlur={[Function onBlur]}
      onFocus={[Function onFocus]}
      placeholder=\\"\\"
      placeholderTextColor=\\"#626262\\"
      rejectResponderTermination={true}
-     secureTextEntry={false}
+     secureTextEntry={true}
      style={
        Array [
          Object {
            \\"color\\": \\"#151515\\",
            \\"flexBasis\\": 0,
@@ -81,14 +81,14 @@
      }
      testID=\\"Basculer l'affichage du mot de passe\\"
    >
      <View
        height=\\"100%\\"
-       testID=\\"eye\\"
+       testID=\\"eye-slash\\"
        width=\\"100%\\"
      >
        <Text>
-         eye-SVG-Mock
+         eye-slash-SVG-Mock
        </Text>
      </View>
    </View>
  </View>"
`;

exports[`<PasswordInput /> should render correctly when password is hidden after it was displayed 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -33,11 +33,11 @@
      onBlur={[Function onBlur]}
      onFocus={[Function onFocus]}
      placeholder=\\"\\"
      placeholderTextColor=\\"#626262\\"
      rejectResponderTermination={true}
-     secureTextEntry={true}
+     secureTextEntry={false}
      style={
        Array [
          Object {
            \\"color\\": \\"#151515\\",
            \\"flexBasis\\": 0,
@@ -81,14 +81,14 @@
      }
      testID=\\"Basculer l'affichage du mot de passe\\"
    >
      <View
        height=\\"100%\\"
-       testID=\\"eye-slash\\"
+       testID=\\"eye\\"
        width=\\"100%\\"
      >
        <Text>
-         eye-slash-SVG-Mock
+         eye-SVG-Mock
        </Text>
      </View>
    </View>
  </View>"
`;
